hack/verify-license.py

```diff
@@ -0,0 +1,945 @@
+COPYRIGHT_NAMES = {
+    '<YEAR> <COPYRIGHT HOLDER>',
+    '<YEAR> <OWNER>',
+    '[yyyy] [name of copyright owner]',
+    'copyright holder',
+    'the copyright holder',
+}
+
+# Ignore punctuation differences.
+PUNCT_RE = re.compile(r'[][(){},.]+')
+
+
+def find_licenses(root):
+    '''
+    Find all license files by walking from a root.
+
+    Args:
+        root: the directory to search for licenses inside
+    Returns:
+        licenses: a list of LICENSE files to validate.
+        uncovered: a list of directories that are missing LICENSE files.
+    '''
+    licenses = []
+    uncovered = []
+    for dirpath, dirnames, filenames in os.walk(root, followlinks=True):
+        if dirpath in WEIRD_LICENSE_NAMES:
+            path = os.path.join(dirpath, WEIRD_LICENSE_NAMES[dirpath])
+            if os.path.exists(path):
+                licenses.append(path)
+                dirnames[:] = []  # don't walk further
+                continue
+        for filename in filenames:
+            base, ext = os.path.splitext(filename)
+            if base.lower() in ('licence', 'license') \
+                    and ext in ('', '.txt', '.md', '.code'):
+                licenses.append(os.path.join(dirpath, filename))
+                dirnames[:] = []  # don't walk further
+                break
+        else:  # no license found
+            for filename in filenames:
+                if filename in ('BUILD', 'OWNERS'):
+                    continue  # ignore BUILD files
+                fpath = os.path.join(dirpath, filename)
+                if any(fnmatch.fnmatch(fpath, pat) for pat in IGNORE_PATTERNS):
+                    continue
+                uncovered.append(dirpath)
+                dirnames[:] = []  # don't walk further
+                break
+    licenses.sort()
+    uncovered.sort()
+    return licenses, uncovered
+
+
+def extract_substitutions(diff):
+    '''
+    Given a diff (output by difflib.ndiff), return a sequence of
+    substitutions it represents.
+
+    Args:
+        diff: a list like ['- abcd', '+ abce', '?    ^\n', '  unchanged']
+    Returns:
+        [(removed_string, inserted_string), ...]
+    '''
+    in_sub = False
+    subs = []
+    removed = []
+    added = []
+    for line in diff:
+        op = line[0]
+        token = line[2:]
+        if op == ' ':
+            if in_sub:
+                in_sub = False
+                subs.append((' '.join(removed), ' '.join(added)))
+                removed = []
+                added = []
+        elif op == '+':
+            in_sub = True
+            added.append(token)
+        elif op == '-':
+            in_sub = True
+            removed.append(token)
+    if in_sub:
+        subs.append((' '.join(removed), ' '.join(added)))
+    return subs
+
+
+class UnknownLicenseError(Exception):
+    pass
+
+
+def validate_license(path):
+    '''
+    Given a path to a license, parse it and return information about it.
```

Comment:
Will you better document the mechanics of how this works?

Worth putting the edit_dist part into its own function?

Reply:
done